google/cloud/storage/async_client.h

```diff
@@ -152,6 +154,43 @@ class AsyncClient {
          /*.options=*/std::move(options)});
   }
 
+  /**
+   * Reads the contents of an object.
+   *
+   * When satisfied, the returned future has a reader to asynchronously download
+   * the contents of the given object.
+   *
+   * @param bucket_name the name of the bucket that contains the object.
+   * @param object_name the name of the object to be read.
+   * @param args a list of optional query parameters and/or request headers.
+   *     Valid types for this operation include `DisableCrc32cChecksum`,
+   *     `DisableMD5Hash`, `EncryptionKey`, `Generation`, `IfGenerationMatch`,
+   *     `IfGenerationNotMatch`, `IfMetagenerationMatch`,
+   *     `IfMetagenerationNotMatch`, `UserProject`, and `AcceptEncoding`.
+   *
+   * @par Idempotency
+   * This is a read-only operation and is always idempotent.
+   *
+   * @par Example
+   * @snippet storage_async_samples.cc read-object
+   */
+  template <typename... Args>
+  future<StatusOr<std::pair<AsyncReader, AsyncToken>>> ReadObject(
```

Comment:
How do you pass in args?

With ReadObjectRange() i've tried and failed:

```
namespace gcs = google::cloud::storage;
auto options = google::cloud::Options{}.set<gcs::Generation>(0);

client.ReadObject(bucket_name, object_name, 0, 100, options).get();
```

failed with error:

```
google/cloud/storage/examples/storage_async_samples.cc:180:67: error: no matching function for call to 'google::cloud::v2_16::Options::set<google::cloud::storage::v2_16::Generation>(int)'
  180 |       auto options = google::cloud::Options{}.set<gcs::Generation>(0);
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
In file included from ./google/cloud/internal/call_context.h:18,
                 from ./google/cloud/internal/async_read_stream_impl.h:19,
                 from ./google/cloud/completion_queue.h:19,
                 from ./google/cloud/storage/internal/async/connection.h:21,
                 from ./google/cloud/storage/async_client.h:18,
                 from google/cloud/storage/examples/storage_async_samples.cc:19:
./google/cloud/options.h:139:12: note: candidate: 'template<class T> google::cloud::v2_16::Options& google::cloud::v2_16::Options::set(ValueTypeT<T>) &'
  139 |   Options& set(ValueTypeT<T> v) & {
      |            ^~~
./google/cloud/options.h:139:12: note:   template argument deduction/substitution failed:
./google/cloud/options.h: In substitution of 'template<class T> using ValueTypeT = typename T::Type [with T = google::cloud::storage::v2_16::Generation]':
./google/cloud/options.h:139:12:   required by substitution of 'template<class T> google::cloud::v2_16::Options& google::cloud::v2_16::Options::set(ValueTypeT<T>) & [with T = google::cloud::storage::v2_16::Generation]'
google/cloud/storage/examples/storage_async_samples.cc:180:67:   required from here
./google/cloud/options.h:97:9: error: no type named 'Type' in 'struct google::cloud::storage::v2_16::Generation'
   97 |   using ValueTypeT = typename T::Type;
      |         ^~~~~~~~~~
./google/cloud/options.h:158:13: note: candidate: 'template<class T> google::cloud::v2_16::Options&& google::cloud::v2_16::Options::set(ValueTypeT<T>) &&'
  158 |   Options&& set(ValueTypeT<T> v) && {
      |             ^~~
      ```

Reply:
`client.ReadObject(bucket_name, object_name, gcs::Generation(12345));`

Reply:
gotcha; i tried that as well and see:

```
./google/cloud/storage/async_client.h: In instantiation of 'google::cloud::v2_16::future<google::cloud::storage_experimental::v2_16::AsyncReadObjectRangeResponse> google::cloud::storage_experimental::v2_16::AsyncClient::ReadObject(const std::string&, const std::string&, int64_t, int64_t, RequestOptions&& ...) [with RequestOptions = {google::cloud::storage::v2_16::Generation}; std::string = std::__cxx11::basic_string<char>; int64_t = long int]':
google/cloud/storage/examples/storage_async_samples.cc:181:24:   required from here
./google/cloud/storage/async_client.h:206:42: error: no matching function for call to 'forward<google::cloud::v2_16::Options>(google::cloud::storage::v2_16::Generation&)'
  206 |         SpanOptions(std::forward<Options>(options)...));
      |                     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
```

Reply:
I just wrote an additional example and it compiled for me.  Give me 10m, I will merge this PR and send a second PR with the additional example.